### ISSUE TITLE
sort manifests output by 'helm install --dry-run --debug'

### ISSUE THREAD

**Issue by sbezverk** [kubernetes/helm#1696]
Noticed a very strange behavior when rendering the same chart (parent chart with sub charts) shows different results. Basically I just do Arrow UP and Enter and  without any modifications this renders differently:
! Working case
kind: PersistentVolumeClaim
metadata:
  name: glance-test
! Incorrectly working case
kind: PersistentVolumeClaim
metadata:
  name:   

I managed to collect strace for working and non working case and I see in non-working case common macro file located in parent chart's template folder does not get opened, when in the working case you can clearly see it gets opened.

Attaching the complete strace log for both cases for review.
[changing values log dec 16 2016.txt](https://github.com/kubernetes/helm/files/657154/changing.values.log.dec.16.2016.txt)

**Comment by technosophos** [comment]
Was this just a case of two helper templates being called? Is this still a problem?

**Comment by sbezverk** [comment]
looks clean so far, closing it..

**Comment by goruha** [comment]
Bug is still there

We expect that two consecutive runs would have essentially the same output (~almost idempotent), with exception of dynamically generated fields (e.g. passwords and dates).

What we observe is that this is not the case. the order of the resources are entirely different. this has the negative consequence: `_helpers.yaml` in umbrella charts which may define blocks might not be called first.

The expectation is that `_helpers.yaml` would always be rendered first in umbrella charts

But we have 

```sh
helm install --name=lamp2 -f lamp/.values.yaml lamp --dry-run --debug   > /tmp/v1.yaml
helm install --name=lamp2 -f lamp/.values.yaml lamp --dry-run --debug   > /tmp/v2.yaml
diff /tmp/v1.yaml /tmp/v2.yaml
```

returns output

````
--- /tmp/v1.yaml
+++ /tmp/v2.yaml
@@ -1,9 +1,9 @@
-Created tunnel using local port: '34061'
-SERVER: "localhost:34061"
+Created tunnel using local port: '37135'
+SERVER: "localhost:37135"
 CHART PATH: /home/goruha/projects/cloudposse/charts/incubator/lamp
 NAME:   lamp2
 REVISION: 1
-RELEASED: Thu Feb  9 20:01:28 2017
+RELEASED: Thu Feb  9 20:01:35 2017
 CHART: lamp-0.1.0
 USER-SUPPLIED VALUES:
 apache:
@@ -248,7 +248,7 @@
 
 HOOKS:
 ---
-# lamp2-lamp-ksesbhzifl
+# lamp2-lamp-a4eqqwlerf
 apiVersion: batch/v1
 kind: Job
 metadata:
@@ -259,7 +259,7 @@
     chart: lamp-0.1.0
     heritage: Tiller
     release: lamp2
-  name: lamp2-lamp-ksesbhzifl
+  name: lamp2-lamp-a4eqqwlerf
 spec:
   template:
     metadata:
@@ -323,7 +323,7 @@
     chart: "vps-0.1.0"
 data:
   login: "YWRtaW4="
-  password: "SFdjelF5OWhRRw=="
+  password: "ek9raFc5NldTMg=="
 ---
 # Source: lamp/charts/mysql/templates/secrets.yaml
 apiVersion: v1
@@ -338,7 +338,7 @@
 type: Opaque
 data:
   
-  mysql-root-password: "N3NBeFdkRk1Sag=="
+  mysql-root-password: "QnAxUkZhZlk0ZQ=="
   
   
   mysql-password:  "d3AtcGFzc3dvcmQ="
@@ -359,24 +359,22 @@
     define('DB_HOST', $MYSQL_DB_HOST);
     define('DOMAIN_CURRENT_SITE', '');
 ---
-# Source: lamp/charts/mysql/templates/svc.yaml
+# Source: lamp/charts/vps/templates/service.yaml
 apiVersion: v1
 kind: Service
 metadata:
-  name: lamp2-mysql
+  name: lamp2-vps
   labels:
-    app: lamp2-mysql
-    chart: "mysql-0.2.4"
-    release: "lamp2"
-    heritage: "Tiller"
+    chart: "vps-0.1.0"
 spec:
+  type: ClusterIP
   ports:
-  - name: mysql
-    port: 3306
-    targetPort: mysql
+  - port: 22
+    targetPort: 22
+    protocol: TCP
+    name: ssh
   selector:
-    app: lamp2-mysql
-
+    app: lamp2-vps
 ---
 # Source: lamp/charts/apache/templates/service.yaml
 apiVersion: v1
@@ -396,27 +394,10 @@
     app: lamp2-apache
 
 ---
-# Source: lamp/charts/vps/templates/service.yaml
+# Source: lamp/charts/mysql/templates/svc.yaml
 apiVersion: v1
 kind: Service
 metadata:
-  name: lamp2-vps
-  labels:
-    chart: "vps-0.1.0"
-spec:
-  type: ClusterIP
-  ports:
-  - port: 22
-    targetPort: 22
-    protocol: TCP
-    name: ssh
-  selector:
-    app: lamp2-vps
----
-# Source: lamp/charts/mysql/templates/deployment.yaml
-apiVersion: extensions/v1beta1
-kind: Deployment
-metadata:
   name: lamp2-mysql
   labels:
     app: lamp2-mysql
@@ -424,136 +405,12 @@
     release: "lamp2"
     heritage: "Tiller"
 spec:
-  template:
-    metadata:
-      labels:
-        app: lamp2-mysql
-      annotations:
-        pod.beta.kubernetes.io/init-containers: '[
-              {
-                  "name": "remove-lost-found",
-                  "image": "busybox:1.25.0",
-                  "command": ["rm", "-fr", "/var/lib/mysql/lost+found"],
-                  "volumeMounts": [
-                      {
-                          "name": "data",
-                          "mountPath": "/var/lib/mysql"
-                      }
-                  ],
-                  "imagePullPolicy": "IfNotPresent"
-              }
-          ]'
-    spec:
-      containers:
-      - name: lamp2-mysql
-        image: "mysql:5.7.14"
-        imagePullPolicy: "IfNotPresent"
-        resources:
-          requests:
-            cpu: 100m
-            memory: 256Mi
-          
-        env:
-        - name: MYSQL_ROOT_PASSWORD
-          valueFrom:
-            secretKeyRef:
-              name: lamp2-mysql
-              key: mysql-root-password
-        - name: MYSQL_PASSWORD
-          valueFrom:
-            secretKeyRef:
-              name: lamp2-mysql
-              key: mysql-password
-        - name: MYSQL_USER
-          value: "wp"
-        - name: MYSQL_DATABASE
-          value: "wp"
-        ports:
-        - name: mysql
-          containerPort: 3306
-        livenessProbe:
-          exec:
-            command:
-            - mysqladmin
-            - ping
-          initialDelaySeconds: 30
-          timeoutSeconds: 5
-        readinessProbe:
-          exec:
-            command:
-            - mysqladmin
-            - ping
-          initialDelaySeconds: 5
-          timeoutSeconds: 1
-        volumeMounts:
-        - name: data
-          mountPath: /var/lib/mysql
-      volumes:
-      - name: data
-        persistentVolumeClaim:
-          claimName: lamp2-mysql
----
-# Source: lamp/charts/vps/templates/deployment.yaml
-apiVersion: extensions/v1beta1
-kind: Deployment
-metadata:
-  name: lamp2-vps
-  labels:
-    chart: "vps-0.1.0"
-spec:
-  replicas: 1
-  template:
-    metadata:
-      labels:
-        app: lamp2-vps
-      annotations:
-        checksum/config: 373dbec729fdbe710e14601dec0f41c68bafb5e764981ab200c8789a526034a1
-    spec:
-      containers:
-      - name: vps
-        image: "cloudposse/ubuntu-vps:latest"
-        imagePullPolicy: Always
-        env:
-        - name: VPS_PASSWORD
-          valueFrom:
-            secretKeyRef:
-              name: lamp2-vps
-              key: password
-        - name: VPS_USER
-          valueFrom:
-            secretKeyRef:
-              name: lamp2-vps
-              key: login
-        - name: VPS_GROUP
-          value: "admin"
-        - name: VPS_ENABLE_SUDO
-          value: "true"
-        - name: VPS_GITHUB_USERS
-          value: 
-        ports:
-        - containerPort: 22
-        livenessProbe:
-          tcpSocket:
-            port: 22
-        readinessProbe:
-          tcpSocket:
-            port: 22
-        resources:
-            limits:
-              cpu: 100m
-              memory: 128Mi
-            requests:
-              cpu: 100m
-              memory: 128Mi
-            
-        volumeMounts:
-          - mountPath: "/data"
-            name: "data"
-      volumes:
-        - name: "data"
-          persistentVolumeClaim:
-            claimName: wordpress-test
-          
+  ports:
+  - name: mysql
+    port: 3306
+    targetPort: mysql
+  selector:
+    app: lamp2-mysql
 
 ---
 # Source: lamp/charts/apache/templates/deployment.yaml
@@ -657,22 +514,149 @@
             claimName: wordpress-test
           
 ---
-# Source: lamp/templates/pvc.yaml
-kind: PersistentVolumeClaim
-apiVersion: v1
+# Source: lamp/charts/vps/templates/deployment.yaml
+apiVersion: extensions/v1beta1
+kind: Deployment
 metadata:
-  name: "wordpress-test"
+  name: lamp2-vps
   labels:
-    chart: "lamp-0.1.0"
-  annotations:
-    volume.beta.kubernetes.io/storage-class: "local-nfs"
+    chart: "vps-0.1.0"
 spec:
-  accessModes:
-    - ReadWriteMany
-  resources:
-    requests:
-      storage: "1Gi"
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        app: lamp2-vps
+      annotations:
+        checksum/config: 373dbec729fdbe710e14601dec0f41c68bafb5e764981ab200c8789a526034a1
+    spec:
+      containers:
+      - name: vps
+        image: "cloudposse/ubuntu-vps:latest"
+        imagePullPolicy: Always
+        env:
+        - name: VPS_PASSWORD
+          valueFrom:
+            secretKeyRef:
+              name: lamp2-vps
+              key: password
+        - name: VPS_USER
+          valueFrom:
+            secretKeyRef:
+              name: lamp2-vps
+              key: login
+        - name: VPS_GROUP
+          value: "admin"
+        - name: VPS_ENABLE_SUDO
+          value: "true"
+        - name: VPS_GITHUB_USERS
+          value: 
+        ports:
+        - containerPort: 22
+        livenessProbe:
+          tcpSocket:
+            port: 22
+        readinessProbe:
+          tcpSocket:
+            port: 22
+        resources:
+            limits:
+              cpu: 100m
+              memory: 128Mi
+            requests:
+              cpu: 100m
+              memory: 128Mi
+            
+        volumeMounts:
+          - mountPath: "/data"
+            name: "data"
+      volumes:
+        - name: "data"
+          persistentVolumeClaim:
+            claimName: wordpress-test
+          
+
 ---
+# Source: lamp/charts/mysql/templates/deployment.yaml
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: lamp2-mysql
+  labels:
+    app: lamp2-mysql
+    chart: "mysql-0.2.4"
+    release: "lamp2"
+    heritage: "Tiller"
+spec:
+  template:
+    metadata:
+      labels:
+        app: lamp2-mysql
+      annotations:
+        pod.beta.kubernetes.io/init-containers: '[
+              {
+                  "name": "remove-lost-found",
+                  "image": "busybox:1.25.0",
+                  "command": ["rm", "-fr", "/var/lib/mysql/lost+found"],
+                  "volumeMounts": [
+                      {
+                          "name": "data",
+                          "mountPath": "/var/lib/mysql"
+                      }
+                  ],
+                  "imagePullPolicy": "IfNotPresent"
+              }
+          ]'
+    spec:
+      containers:
+      - name: lamp2-mysql
+        image: "mysql:5.7.14"
+        imagePullPolicy: "IfNotPresent"
+        resources:
+          requests:
+            cpu: 100m
+            memory: 256Mi
+          
+        env:
+        - name: MYSQL_ROOT_PASSWORD
+          valueFrom:
+            secretKeyRef:
+              name: lamp2-mysql
+              key: mysql-root-password
+        - name: MYSQL_PASSWORD
+          valueFrom:
+            secretKeyRef:
+              name: lamp2-mysql
+              key: mysql-password
+        - name: MYSQL_USER
+          value: "wp"
+        - name: MYSQL_DATABASE
+          value: "wp"
+        ports:
+        - name: mysql
+          containerPort: 3306
+        livenessProbe:
+          exec:
+            command:
+            - mysqladmin
+            - ping
+          initialDelaySeconds: 30
+          timeoutSeconds: 5
+        readinessProbe:
+          exec:
+            command:
+            - mysqladmin
+            - ping
+          initialDelaySeconds: 5
+          timeoutSeconds: 1
+        volumeMounts:
+        - name: data
+          mountPath: /var/lib/mysql
+      volumes:
+      - name: data
+        persistentVolumeClaim:
+          claimName: lamp2-mysql
+---
 # Source: lamp/charts/mysql/templates/pvc.yaml
 
 kind: PersistentVolumeClaim
@@ -693,3 +677,19 @@
     requests:
       storage: "8Gi"
 
+---
+# Source: lamp/templates/pvc.yaml
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: "wordpress-test"
+  labels:
+    chart: "lamp-0.1.0"
+  annotations:
+    volume.beta.kubernetes.io/storage-class: "local-nfs"
+spec:
+  accessModes:
+    - ReadWriteMany
+  resources:
+    requests:
+      storage: "1Gi"
````

**Comment by goruha** [comment]
@sbezverk please reopen this issue

**Comment by goruha** [comment]
The first error should be here
https://github.com/kubernetes/helm/blob/c1e6683233537c24c99b3228cb690c18c6d0f9c7/pkg/tiller/release_server.go#L750

```go
	for k, v := range files {
		if strings.HasSuffix(k, notesFileSuffix) {
			// Only apply the notes if it belongs to the parent chart
			// Note: Do not use filePath.Join since it creates a path with \ which is not expected
			if k == path.Join(ch.Metadata.Name, "templates", notesFileSuffix) {
				notes = v
			}
			delete(files, k)
		}
	}
```
http://stackoverflow.com/questions/9619479/go-what-determines-the-iteration-order-for-map-keys
https://blog.golang.org/go-maps-in-action

Right workaround

```golang
import "sort"

var m map[int]string
var keys []int
for k := range m {
    keys = append(keys, k)
}
sort.Ints(keys)
for _, k := range keys {
    fmt.Println("Key:", k, "Value:", m[k])
}
```


The second must have behaviour is to have templates started with _ be rendered the first order

**Comment by sbezverk** [comment]
as per goruha''s request reopen the issue

**Comment by technosophos** [comment]
Hmmm... we've never guaranteed that ordering would remain the same on `--debug` output, but I'd accept a patch if someone commits it.

Since manifests get sorted elsewhere later in the install process (by resource type), any sort operation that happens _specifically for debug_ should be wrapped appropriately.

It's possible that we inadvertently fixed this in 2.2.0, since we changed the datastructure for manifest files. But I haven't tested that at all.
